README.md
[![pub package](https://img.shields.io/pub/v/flutter_sortable_wrap.svg)](https://pub.dev/packages/flutter_sortable_wrap)

##### Build & Run `example/lib/main.dart` on iOS/Android/Chrome/MacOS for more demonstrations
`cd example/; flutter clean; flutter run -d macOS`


#### Simple to use

```
SortableWrap(
  children: children,
  onSorted: (int oldIndex, int newIndex) {
    setState(() {
      your_data_array.insert(newIndex, your_data_array.removeAt(oldIndex));
      print('----->>>>> your_data_array: $your_data_array');
    });
  },
  spacing: 10,
  runSpacing: 15,
)
```


### Demonstrations

<img src="https://raw.githubusercontent.com/isaacselement/flutter_sortable_wrap/master/example/resources/Kapture%202023-03-22%20at%2017.59.11.gif" width="32%">


##### Features and Issues

Please feel free to request new features and bugs at the [issue tracker][tracker]

[tracker]: https://github.com/isaacselement/flutter_sortable_wrap/issues

CHANGELOG.md

# 1.0.5

* Support long press then become draggable.






lib/sortable_item.dart
// ignore_for_file: constant_identifier_names

import 'package:flutter/material.dart';
import 'package:flutter_sortable_wrap/sortable_wrap.dart';

class SortableItem extends StatefulWidget {
  SortableItem({
    Key? key,
    required this.element,
    required this.onEventHit,
  }) : super(key: key) {
    element.widget = this;
  }

  final OnEventHit onEventHit;
  final SortableElement element;

  @override
  SortableItemState createState() => SortableItemState();
}

class SortableItemState extends State<SortableItem> with TickerProviderStateMixin {
  /// Animations controllers
  late AnimationController slideToRightController;
  late AnimationController slideToLeftController;

  /// Use for creating a clone ghost view when user dragging to different row
  GhostType ghostType = GhostType.None;

  void setGhostType(GhostType type) => setState(() => ghostType = type);

  /// Visible index for rolling from / rolling to
  late int sourceIndex;
  late int destinationIndex;

  /// Start the rolling animation
  void startAnimation(bool isDraggingInSameRow) {
    bool isSlideToRight = destinationIndex > sourceIndex;
    bool isSlideToLeft = destinationIndex < sourceIndex;
    TickerFuture? animationFuture;
    if (isSlideToRight) {
      /// slide to right, use 'reverse', offset.x: -1.0 -> 0.0
      animationFuture = slideToLeftController.reverse(from: 1.0);
    } else if (isSlideToLeft) {
      /// slide to left, use 'reverse', offset.x: 1.0 -> 0.0
      animationFuture = slideToRightController.reverse(from: 1.0);
    }

    /// dragging in the same row/line, just return
    if (isDraggingInSameRow) return;

    /// TODO ... Hit the first/last one, in this case animation effect issue ...
    /// TODO ... Dragging one is the first/last one, in this case animation effect issue

    /// stick a ghost for the last/first position element
    if (isSlideToRight) {
      if (element.isTheLastOne) {
        setGhostType(GhostType.Next);
      }
    } else if (isSlideToLeft) {
      if (element.isTheFirstOne) {
        setGhostType(GhostType.Previous);
      }
    }
    if (ghostType != GhostType.None) {
      animationFuture?.then((value) {
        setGhostType(GhostType.None);
      });
    }
  }

  /// Data or Relation holder, anyway ...
  SortableElement get element => widget.element;

  @override
  void initState() {
    super.initState();
    slideToRightController = AnimationController(vsync: this, duration: const Duration(milliseconds: 200));
    slideToLeftController = AnimationController(vsync: this, duration: const Duration(milliseconds: 200));
  }

  @override
  void dispose() {
    /// call animation controller first before super.dispose()
    slideToRightController.dispose();
    slideToLeftController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    element.state = this;

    /// append capacity of animation
    Widget _capacityAnimate(Widget child) {
      child = SlideTransition(
        child: child,
        position:
            Tween<Offset>(begin: const Offset(0.0, 0.0), end: const Offset(1.0, 0.0)).animate(slideToRightController),
      );
      child = SlideTransition(
        child: child,
        position:
            Tween<Offset>(begin: const Offset(0.0, 0.0), end: const Offset(-1.0, 0.0)).animate(slideToLeftController),
      );
      return child;
    }

    /// append capacity of able be hit
    Widget _capacityHit(Widget child, OnEventHit onEventHit) {
      List<Widget> children = [
        child,
        Positioned(
          top: 0,
          left: 0,
          right: 0,
          bottom: 0,
          child: DragTarget<SortableElement>(
            builder: (BuildContext context, List<SortableElement?> candidateData, List<dynamic> rejectedData) {
              return const SizedBox();
              // return ColoredBox(color: Colors.grey.withAlpha(128), child: SizedBox());  // for debug :)
            },
            onWillAccept: (SortableElement? toAccept) {
              if (toAccept == null) return false;
              onEventHit(this, toAccept);
              return true;
            },
            onAccept: (SortableElement accepted) {},
            onLeave: (Object? leaving) {},
          ),
        ),
      ];

      /// insert or append a ghost/clone view
      if (ghostType != GhostType.None) {
        double spacing = element.parent.widget.spacing;
        double width = element.parent.anyElementSize.width;
        bool isFrontGhost = ghostType == GhostType.Previous;
        double left = isFrontGhost ? -width - spacing : width + spacing;
        SortableElement sibling = isFrontGhost ? element.previousToMe : element.nextToMe;
        children.insert(
            isFrontGhost ? 0 : children.length, Positioned(top: 0, bottom: 0, left: left, child: sibling.view));
      }
      return Stack(clipBehavior: Clip.none, children: children);
    }

    return _capacityAnimate(_capacityHit(element.view, widget.onEventHit));
  }
}

/// Data or Relation Model
class SortableElement {
  /// The caller's view
  late Widget view;

  /// The caller's children index, the most original index
  late int originalIndex;

  /// The [SortableWrapState] context/state i'm staying in
  late SortableWrapState parent;

  /// Element index before dragging start
  late int preservedIndex;

  /// The [SortableItem] widget i'm binding to
  late SortableItem widget;

  /// The [SortableItemState] widget i'm binding to, corresponding to this [widget]
  late SortableItemState state;

  /// Element index of on rolling what you are looking at
  int get visibleIndex => parent.animationElements.indexOf(this);

  SortableElement get nextToMe => parent.animationElements[visibleIndex + 1];

  SortableElement get previousToMe => parent.animationElements[visibleIndex - 1];

  /// If is the first one on a row
  bool get isTheFirstOne => isTheFirstOnRow(parent.animationElements);

  /// If is the last one on a row
  bool get isTheLastOne => isTheLastOnRow(parent.animationElements);

  bool isTheFirstOnRow(List<SortableElement> elements) {
    int index = elements.indexOf(this);
    return index != 0 && index % parent.elementCountPerRow == 0;
  }

  bool isTheLastOnRow(List<SortableElement> elements) {
    int index = elements.indexOf(this);
    return index != elements.length - 1 && (index + 1) % parent.elementCountPerRow == 0;
  }
}

enum GhostType { None, Previous, Next }

typedef OnEventHit = void Function(SortableItemState state, SortableElement accept);


lib/sortable_wrap.dart
import 'package:flutter/material.dart';
import 'package:flutter_sortable_wrap/sortable_item.dart';
import 'package:flutter_sortable_wrap/sortable_utils.dart';

class SortableWrap extends StatefulWidget {
  const SortableWrap({
    Key? key,
    required this.children,
    required this.onSorted,
    this.onSortStart,
    this.onSortCancel,
    this.spacing = 0.0,
    this.runSpacing = 0.0,
    this.options,
  }) : super(key: key);

  final List<Widget> children;

  /// Events
  final void Function(int oldIndex, int newIndex) onSorted;
  final void Function(int index)? onSortStart;
  final void Function(int index)? onSortCancel;

  /// TODO ... To complete the remaining properties that need to pass to [Wrap] ...
  /// Properties pass to the official [Wrap] widget
  final double spacing;
  final double runSpacing;

  /// Widget settings
  final SortableWrapOptions? options;

  @override
  State<SortableWrap> createState() => SortableWrapState();
}

class SortableWrapState extends State<SortableWrap> {
  /// BuildContexts & Size & Count Properties, use them after determined
  late BuildContext wrapperContext;
  late BuildContext anyElementContext;
  late Size wrapperSize;
  late Size anyElementSize;
  late int elementCountPerRow;
  late int elementCountPerColumn;

  /// Dragging & Index Properties
  SortableElement? draggingElement;

  bool get isDragging => draggingElement != null;

  bool isDraggingMe(SortableElement e) => draggingElement == e;

  /// Cached array that keep the index status before swap on rolling
  List<SortableElement> preservedElements = [];

  /// Cached array that representing the realtime swap index when a drag is under way
  List<SortableElement> animationElements = [];

  /// Widget settings
  SortableWrapOptions? _options;

  SortableWrapOptions get options => widget.options ?? (_options ??= SortableWrapOptions());

  @override
  void initState() {
    super.initState();
    initCachedWithChildren();
  }

  @override
  void didUpdateWidget(covariant SortableWrap oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.children.length != oldWidget.children.length) {
      /// TODO ... enhance when dragging ...
      initCachedWithChildren();
    }
  }

  void initCachedWithChildren() {
    preservedElements.clear();
    for (int i = 0; i < widget.children.length; i++) {
      SortableElement e = SortableElement();
      e.view = widget.children[i];
      e.originalIndex = i;
      e.parent = this;
      preservedElements.add(e);
    }
    syncPreservedCacheIndexes();

    animationElements.clear();
    animationElements.addAll(preservedElements);
  }

  void syncPreservedCacheIndexes() {
    for (int i = 0; i < preservedElements.length; i++) {
      preservedElements[i].preservedIndex = i;
    }
  }

  @override
  void dispose() {
    super.dispose();

    /// clear cached
    draggingElement = null;
    preservedElements.clear();
    animationElements.clear();
  }

  @override
  Widget build(BuildContext context) {
    Widget builder = Builder(
      builder: (context) {
        wrapperContext = context;
        return Wrap(
          spacing: widget.spacing,
          runSpacing: widget.runSpacing,
          children: animationElements.map((e) => enclosedWithDraggable(e)).toList(),
        );
      },
    );
    // return ColoredBox(color: Colors.grey, child: builder); // for debug :)
    Widget clip(Widget child) => ClipRect(child: child);
    return clip(builder);
  }

  /// Wrapped with draggable widget
  Widget enclosedWithDraggable(SortableElement element) {
    int index = preservedElements.indexOf(element);

    /// A. Return the widget in dragging mode
    if (isDragging) {
      if (isDraggingMe(element)) {
        return IgnorePointer(ignoring: true, child: Opacity(opacity: 0.2, child: element.view));
      } else {
        return SortableItem(key: ValueKey(index), element: element, onEventHit: eventDoRollingInDragging);
      }
    }

    /// B. Return the widget in idle mode
    void safeInvoke(VoidCallback fn) {
      try {
        fn();
      } catch (e, s) {
        iDebugLog('❗👠🩸 ERROR! please check it out: $e, $s');
      }
    }

    /// Drag finish (end/complete/canceled) callback
    void onDragFinished() {
      int newIndex = element.visibleIndex;
      int oldIndex = element.preservedIndex;
      setState(() {
        /// synchronize cache preserved elements with animation elements
        preservedElements.clear();
        preservedElements.addAll(animationElements);
        syncPreservedCacheIndexes();

        draggingElement = null;
      });

      /// invoke caller's callback
      safeInvoke(() {
        if (oldIndex == newIndex) {
          widget.onSortCancel?.call(oldIndex);
        } else {
          widget.onSorted(oldIndex, newIndex);
        }
      });
    }

    /// Drag started callback
    void onDragStarted() {
      iDebugLog('Drag start index: $index');
      iDebugLog('Context\'s size: ${wrapperContext.size}');
      iDebugLog('Element\'s size: ${anyElementContext.size}');

      /// calculate the row & column & count/size at this moment, for window may resizeable when running as Desktop App
      assert(wrapperContext.size != null, 'Wrap\'s size cannot be determined!');
      assert(anyElementContext.size != null, 'Element\'s size cannot be determined!');
      wrapperSize = wrapperContext.size!;
      anyElementSize = anyElementContext.size!;
      double width = wrapperSize.width;
      double height = wrapperSize.height;
      double ew = anyElementSize.width;
      double eh = anyElementSize.height;
      elementCountPerRow = width ~/ ew;
      elementCountPerColumn = height ~/ eh;
      iDebugLog('[START] count on per row: $elementCountPerRow, count on per column: $elementCountPerColumn');
      while (elementCountPerRow * ew + (elementCountPerRow - 1) * widget.spacing > width) {
        elementCountPerRow--;
      }
      while (elementCountPerColumn * eh + (elementCountPerColumn - 1) * widget.runSpacing > height) {
        elementCountPerColumn--;
      }
      iDebugLog('[FINAL] count on per row: $elementCountPerRow, count on per column: $elementCountPerColumn');

      /// set current dragging element
      setState(() {
        draggingElement = element;
      });

      /// invoke caller's callback
      safeInvoke(() => widget.onSortStart?.call(element.preservedIndex));
    }

    /// Drag end callback
    void onDragEnd(DraggableDetails details) {
      onDragFinished();
    }

    /// Drag complete callback
    void onDragCompleted() {
      onDragFinished();
    }

    /// Drag canceled callback
    void onDraggableCanceled(Velocity velocity, Offset offset) {
      onDragFinished();
    }

    /// a key is needed, for keeping DraggableState when inner setState called.
    ValueKey valueKey = ValueKey(index);
    Widget childBuilder = Builder(builder: (context) {
      /// cache a context for calculate size on starting drag
      anyElementContext = context;
      return options.draggableChildBuilder?.call(element.view) ?? element.view;
    });
    return options.isLongPressDraggable
        ? LongPressDraggable<SortableElement>(
            /// the data passing to DragTarget and its callbacks
            data: element,
            key: valueKey,
            child: childBuilder,
            feedback: options.draggableFeedbackBuilder(element.view),
            onDragEnd: onDragEnd,
            onDragStarted: onDragStarted,
            onDragCompleted: onDragCompleted,
            onDraggableCanceled: onDraggableCanceled,
          )
        : Draggable<SortableElement>(
            /// the data passing to DragTarget and its callbacks
            data: element,
            key: valueKey,
            child: childBuilder,
            feedback: options.draggableFeedbackBuilder(element.view),
            onDragEnd: onDragEnd,
            onDragStarted: onDragStarted,
            onDragCompleted: onDragCompleted,
            onDraggableCanceled: onDraggableCanceled,
          );
  }

  /// Events
  void eventDoRollingInDragging(SortableItemState beHitItemState, SortableElement holdingElement) {
    assert(draggingElement != null, 'Dragging status is a mess now, please check it out.');
    assert(draggingElement == holdingElement, 'Got a different dragging view, please check it out.');

    SortableElement dragging = draggingElement!;
    SortableElement element = beHitItemState.widget.element;

    int toIndex = animationElements.indexOf(element);
    int draggingIndex = animationElements.indexOf(dragging);
    bool isDraggingInSameRow = toIndex ~/ elementCountPerRow == draggingIndex ~/ elementCountPerRow;

    /// To lower index means user dragging to left, user dragging to left or top, the hit target should animate to right
    bool isDraggingToLowerIndex = toIndex < draggingIndex;
    int i = isDraggingToLowerIndex ? draggingIndex - 1 : draggingIndex + 1;
    for (; isDraggingToLowerIndex ? i >= toIndex : i <= toIndex; isDraggingToLowerIndex ? i-- : i++) {
      SortableElement e = animationElements[i];

      /// Swap the index in cached data
      int sourceIndex = i;
      int destinationIndex = isDraggingToLowerIndex ? i + 1 : i - 1;
      animationElements.swap(sourceIndex, destinationIndex);

      /// Handle animation by corresponding item's state
      SortableItemState itemState = e.state;
      itemState.sourceIndex = sourceIndex;
      itemState.destinationIndex = destinationIndex;
      itemState.startAnimation(isDraggingInSameRow);
    }

    /// Make sure you see the right thing on the right position
    setState(() {});
  }
}

/// Options for configuring the UI/Animation
class SortableWrapOptions {
  /// should long press and then become draggable
  bool isLongPressDraggable = false;

  Widget Function(Widget child)? draggableChildBuilder;

  /// the dragging moving view builder, typically return a widget wrap by [Material]
  Widget Function(Widget child) draggableFeedbackBuilder = (Widget child) {
    return Material(
      elevation: 18.0,
      child: child,
      color: Colors.transparent,
      shadowColor: Colors.transparent,
      borderRadius: BorderRadius.zero,
    );
  };

  SortableWrapOptions();

  /// clone a new instance
  SortableWrapOptions clone() {
    SortableWrapOptions newInstance = SortableWrapOptions();

    newInstance.draggableFeedbackBuilder = draggableFeedbackBuilder;

    return newInstance;
  }
}


example/lib/page/page_next.dart
import 'package:flutter/material.dart';
import 'package:flutter_sortable_wrap/flutter_sortable_wrap.dart';

class PageNext extends StatefulWidget {
  const PageNext({Key key}) : super(key: key);

  @override
  PageNextState createState() => PageNextState();
}

class PageNextState extends State<PageNext> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: SafeArea(
          child: Container(
            alignment: Alignment.topCenter,
            child: CustomScrollView(
              cacheExtent: 0,
              slivers: [
                SliverPersistentHeader(
                  pinned: true,
                  floating: true,
                  delegate: MyDelegate(),
                ),
              ],
            ),
          ),
        ),
        floatingActionButton: FloatingActionButton(
          child: const Icon(Icons.arrow_back),
          onPressed: () => Navigator.of(context).pop(),
        ));
  }
}

class MyDelegate extends SliverPersistentHeaderDelegate {
  MyDelegate();

  @override
  Widget build(BuildContext context, double shrinkOffset, bool overlapsContent) {
    Widget boxText(String text) => SizedBox(width: 72, height: 72, child: Center(child: Text(text)));

    return SortableWrap(
      onSorted: (int oldIndex, int newIndex) {
        iDebugLog('####### oldIndex: $oldIndex, newIndex: $newIndex');
      },
      children: List.generate(
          10,
          (index) => Padding(
                padding: const EdgeInsets.all(2),
                child: ColoredBox(color: Colors.grey, child: boxText('$index')),
              )).toList(),
    );
  }

  @override
  double get maxExtent {
    return 200;
  }

  @override
  double get minExtent {
    return 200;
  }

  @override
  bool shouldRebuild(covariant MyDelegate oldDelegate) {
    return false;
  }
}


example/lib/main.dart
import 'package:example/page/page_next.dart';
import 'package:flutter/material.dart';
import 'package:flutter_sortable_wrap/flutter_sortable_wrap.dart';

void main() {
  runApp(const MaterialApp(debugShowCheckedModeBanner: false, home: App()));
}

class App extends StatefulWidget {
  const App({Key key}) : super(key: key);

  @override
  AppState createState() => AppState();
}

class AppState extends State<App> {
  List<int> yourDataArray = [];

  @override
  void initState() {
    super.initState();

    yourDataArray.clear();
    for (int i = 0; i < 21; i++) {
      yourDataArray.add(i);
    }
  }

  @override
  Widget build(BuildContext context) {
    iDebugLog('App rebuild!!!!');
    Color alphaColor(Color color, {int alpha = 128}) => color.withAlpha(alpha);
    List<Color> colors = [
      Colors.redAccent,
      Colors.blueAccent,
      Colors.pinkAccent,
      Colors.greenAccent,
      Colors.amberAccent,
      Colors.purpleAccent,
    ];
    colors = colors.map((e) => alphaColor(e)).toList();
    Widget boxText(String text) => SizedBox(width: 72, height: 72, child: Center(child: Text(text)));
    List<Widget> children = [
      for (int i = 0; i < yourDataArray.length; i++)
        ColoredBox(color: colors[yourDataArray[i] % colors.length], child: boxText('${yourDataArray[i]}'))
    ];

    SortableWrapOptions options = SortableWrapOptions();
    options.draggableFeedbackBuilder = (Widget child) {
      return Material(
        elevation: 18.0,
        child: Card(child: child),
        shadowColor: Colors.grey,
        color: Colors.transparent,
        borderRadius: BorderRadius.zero,
      );
    };

    return Scaffold(
      body: SafeArea(
        child: Container(
          alignment: Alignment.topCenter,
          color: Colors.white.withAlpha(64),
          child: SortableWrap(
            children: children,
            onSorted: (int oldIndex, int newIndex) {
              setState(() {
                yourDataArray.insert(newIndex, yourDataArray.removeAt(oldIndex));
                iDebugLog('Data sorted after >>>>>: $yourDataArray');
              });
            },
            onSortStart: (int index) {
              iDebugLog('Data sorted before >>>>>: $yourDataArray');
            },
            spacing: 10,
            runSpacing: 15,
            options: options,
          ),
        ),
      ),
      floatingActionButton: FloatingActionButton(child: const Icon(Icons.add), onPressed: goToNextPage),
    );
  }

  void goToNextPage() {
    Navigator.of(context, rootNavigator: true).push(PageRouteBuilder(
      pageBuilder: (ctx, one, two) => const PageNext(),
      transitionsBuilder: (ctx, one, two, child) => SlideTransition(
        position: Tween<Offset>(begin: const Offset(1.0, 0.0), end: const Offset(0.0, 0.0)).animate(one),
        child: child,
      ),
    ));
  }
}
